src/storage/txn/commands/resolve_lock_scan.rs

```diff
@@ -0,0 +1,145 @@
+            /// `txn_status`. `"k4"` will not be affected either, because it doesn't have a non-committed
+            /// version.
+            txn_status: HashMap<TimeStamp, TimeStamp>,
+            first_scan_key: Option<Key>,
+        }
+}
+
+impl CommandExt for ResolveLockScan {
+    ctx!();
+    tag!(resolve_lock);
+
+    command_method!(readonly, bool, false);
+    command_method!(is_sys_cmd, bool, true);
+
+    fn write_bytes(&self) -> usize {
+        0
+    }
+
+    gen_lock!(first_scan_key: multiple); // it is not multiple actually, but Option is also IntoIterator
+}
+
+impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for ResolveLockScan {
+    fn process_write(self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
+        let (ctx, txn_status) = (self.ctx, self.txn_status);
+        let lock_mgr = context.lock_mgr;
+        let mut reader = MvccReader::new(
+            snapshot.clone(),
+            Some(ScanMode::Forward),
+            !ctx.get_not_fill_cache(),
+            ctx.get_isolation_level(),
+        );
+        let mut txn = MvccTxn::new(
+            snapshot,
+            TimeStamp::zero(),
+            !ctx.get_not_fill_cache(),
+            context.concurrency_manager,
+        );
+        let iter = reader.scan_locks_iter(self.first_scan_key.as_ref())?;
+        let mut rows = 0;
+        let mut next_first_scan_key = None;
+        let mut released_locks = HashMap::default();
+        for item in iter {
+            let (current_key, current_lock) = item?;
+            if !txn_status.contains_key(&current_lock.ts) {
+                continue;
+            }
+            if txn.write_size() >= MAX_TXN_WRITE_SIZE {
+                next_first_scan_key = Some((current_key, current_lock));
+                break;
+            }
+            let mut lock = context.latches.gen_lock(std::iter::once(&current_key));
+            if context.latches.acquire(&mut lock, context.cid) {
+                // when latch is acquired successfully, we can resolve the lock
+                txn.set_start_ts(current_lock.ts);
+                let commit_ts = *txn_status
+                    .get(&current_lock.ts)
+                    .expect("txn status not found");
+
+                let released = if commit_ts.is_zero() {
+                    txn.rollback(current_key.clone())?
+                } else if commit_ts > current_lock.ts {
+                    txn.commit(current_key.clone(), commit_ts)?
+                } else {
+                    return Err(Error::from(ErrorInner::InvalidTxnTso {
+                        start_ts: current_lock.ts,
+                        commit_ts,
+                    }));
+                };
+                released_locks
+                    .entry(current_lock.ts)
+                    .or_insert_with(|| ReleasedLocks::new(current_lock.ts, commit_ts))
+                    .push(released);
+                rows += 1;
+                context.latches.release(&lock, context.cid);
```

Comment:
The latch cannot be released until the mutation is applied. It should be released in `on_write_finished` which is the callback of `async_write`.